### PR TITLE
[RFC] Inherit fields from interface automatically

### PR DESCRIPTION
**Pull request by tobiasdiez**
This is currently a Stage 0: Strawman proposal looking to advance to Stage 1.
The RFC is mainly based on the discussion in the related issues https://github.com/graphql/graphql-spec/issues/533 https://github.com/graphql/graphql-spec/issues/500 and https://github.com/graphql/graphql-js/issues/703.

If someone has additions or clarifications, please feel free to edit the added document directly (if you have the right) or open a PR against my fork. Many thanks in advance!

I hope it is okay that took some of the examples and formulations almost verbatim from the above issues, and expanded on them in form of a RFC. Also the contribution process is not completely clear to me, so please forgive if I overlooked important steps along the way.

### CODE REVIEW

**Comment by leebyron** [comment]
This has been discussed before and rejected. See one comment explaining rationale here https://github.com/graphql/graphql-spec/issues/500#issuecomment-426146328

However I appreciate the work in collecting all details about this RFC in one file. I'm about to move all RFCs to another repository so merging this in advance of doing that

**Comment by tobiasdiez** [comment]
Thanks @leebyron for your reaction. I understand that there are certain objections against automatically inheritance (as for example outlined in your linked comment). But as you see in the reactions to your comment and in #533 (which is under the top 15 of issues with +1), not everyone agrees with the point of favoring readability over not-repeat-yourself. In my opinion, this comes also from the fact that IDE support for graphql caught-up and jumping around the schema is nicely supported right now; in my opinion there is also a shift in using the graphql schema more for writing and using other tools for exploring and reading (like graphql playground, automatically generated docs etc). Due to these natural difference between philosophies, I phrased this RFC as an optional extension - so if a team decides it's more suitable for them to keep the current behavior than that's still possible.

For these reasons I hope the decision to reject this proposal can be revisited. If not, I would suggest to close #533 as well.

**Comment by leebyron** [comment]
You're absolutely welcome to join the next working group meeting. Proposals typically only move forward when their champions bring them to WG meetings to work through the details.

File: rfcs/ImplicitInheritance.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+## Disadvantages
+- It is no longer possible to immediately see all fields that are provided by a type (without traversing the inheritance hierarchy), which makes reading the schema more challenging in some aspects.
+- If a type inherits multiple interfaces that declare a field with different but compatible types, the resulting type is not immediately clear.  
+
+It should be noted that these disadvantages can mostly be mitigated using tools or IDE integrations that can easily determine all fields on the fly.
+
+## Evaluation against [guiding principles](https://github.com/graphql/graphql-spec/blob/main/CONTRIBUTING.md#guiding-principles)
+- Backwards compatibility: ✅
+
+  Existing queries and schema are still valid.
```
Comment (sungam3r) [inline rfcs/ImplicitInheritance.md:192]: Yes, but some GraphQL frameworks/parsers most likely adhere to current "hard" rules about inheriting fields from interfaces so there may be problems when trying to use them with the "new lightweight" schema.

**Comment by meostyles** [comment]
@leebyron This is **absolutely** the worst part about GraphQL, as someone who works on a large type system. Copying fields to many classes is incredibly frustrating:
- you lose any comments on those fields in the interface, so you'd better not forget to copy the comments to each concrete class too!
- developers hack around the limitation with composition, where a class that would otherwise implement an interface and decorate with additional fields instead holds a reference to a concrete type with fields that are shared. This is unfortunate, as it adds an extra layer in the schema since you're introducing a new field, but not entirely unreasonable given the pain of interfaces

This problem was solved decades ago in programming languages. Why is GraphQL stuck in the past?

**Pull request merged**